service/s3/s3crypto/key_provider.go

```diff
@@ -0,0 +1,54 @@
+package s3crypto
+
+import "crypto/rand"
+
+// KeyProvider is used to help generate keys and ivs. This allows for crypto based
+// algorithms or services to be used.
+type KeyProvider interface {
+	EncrypterKeyProvider
+	DecrypterKeyProvider
+}
+
+// EncrypterKeyProvider handles how to generate keys and ivs for the content encryption.
+// This also satifies the `MaterialDescription` and `CipherDataIface` interfaces.
+type EncrypterKeyProvider interface {
```

Comment:
This provider does a lot, what do you think about splitting this interface up? Initially like the generation of key/iv from the get/set methods?  In looking at the usage it doesn't look like the GenerateKey|IV are done in a different context from get/set.